compiler/z/codegen/OMRPeephole.cpp

```diff
@@ -1861,3 +1871,133 @@ OMR::Z::Peephole::tryToRemoveRedundantLTR()
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LLGH, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::LB:
+               if (isSigned)
+                  {
+                  mr = current->getMemoryReference();
+                  mr->resetMemRefUsedBefore();
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LGB, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::LLC:
+               if (isSigned)
+                  {
+                  mr = current->getMemoryReference();
+                  mr->resetMemRefUsedBefore();
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LLGC, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::XR:
+               if (lgfrReg == current->getRegisterOperand(2))
+                  replacement = generateRRInstruction(self()->cg(), TR::InstOpCode::XGR, current->getNode(), lgfrReg, lgfrReg, current->getPrev());
+               break;
+            case TR::InstOpCode::IILF:
+               replacement = generateRILInstruction(self()->cg(), isSigned ? TR::InstOpCode::LGFI : TR::InstOpCode::LLILF, current->getNode(), lgfrReg, toS390RILInstruction(current)->getSourceImmediate(), current->getPrev());
+               break;
+            case TR::InstOpCode::LHI:
+               if (isSigned)
+                  {
+                  replacement = generateRIInstruction(self()->cg(), TR::InstOpCode::LGHI, current->getNode(), lgfrReg, toS390RIInstruction(current)->getSourceImmediate(), current->getPrev());
+                  }
+               else
+                  {
+                  // The following sequence of instructions:
+                  //    LHI      GPR1, IMM   ; sign extend IMM from 16 to 32 bits
+                  //    LLGFR    GPR1, GPR1  ; zero extend from 32 to 64 bits
+                  // Can be converted to
+                  //    LLILF    GPR1, IMM'  ; where IMM' is IMM sign extended from 16 to 32 bits
+                  int16_t imm = toS390RIInstruction(current)->getSourceImmediate();
+                  replacement = generateRILInstruction(self()->cg(), TR::InstOpCode::LLILF, current->getNode(), lgfrReg, static_cast<int32_t>(imm), current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::LR:
+            case TR::InstOpCode::LGR:
+               replacement = generateRRInstruction(self()->cg(), isSigned ? TR::InstOpCode::LGFR : TR::InstOpCode::LLGFR, current->getNode(), lgfrReg, current->getRegisterOperand(2), current->getPrev());
+               break;
+            }
+
+         if (replacement != NULL &&
+            performTransformation(self()->comp(), "O^O S390 PEEPHOLE: Merging %s [%p] and %s [%p] into %s [%p].\n",
```

Comment:
Shouldn't we check performTransformation first even before generating replacement ?